File: ChessProject-PHP/src/Pawn.php
<?php

namespace SolarWinds\Chess;


class Pawn
{
    /** @var PieceColorEnum */
    private $pieceColorEnum;

    /** @var ChessBoard */
    private $chessBoard;

    /** @var int */
    private $xCoordinate;

    /** @var int */
    private $yCoordinate;

    public function __construct(PieceColorEnum $pieceColorEnum)
    {
        $this->pieceColorEnum = $pieceColorEnum;
    }

    public function getChesssBoard()
    {
        return $this->chessBoard;
    }

    public function setChessBoard(ChessBoard $chessBoard)
    {
        $this->chessBoard = $chessBoard;
    }

    /** @return int */
    public function getXCoordinate()
    {
        return $this->xCoordinate;
    }

    /** @var int */
    public function setXCoordinate($value)
    {
        $this->xCoordinate = $value;
    }

    /** @return int */
    public function getYCoordinate()
    {
        return $this->yCoordinate;
    }

    /** @var int */
    public function setYCoordinate($value)
    {
        $this->yCoordinate = $value;
    }

    public function getPieceColor()
    {
        return $this->pieceColorEnum;
    }

    public function setPieceColor(PieceColorEnum $value)
    {
        $this->pieceColorEnum = $value;
    }

    public function move(MovementTypeEnum $movementTypeEnum, $newX, $newY)
    {
        throw new \Exception("Need to implement " . __METHOD__);
    }

    public function toString()
    {
		return "x({$this->xCoordinate}), y({$this->yCoordinate}), pieceColor({$this->pieceColorEnum})";
    }
}

File: ChessProject-Csharp/src/Pawn.cs
using System;

namespace SolarWinds.MSP.Chess
{
    public class Pawn
    {
        private ChessBoard chessBoard;
        private int xCoordinate = -1;
        private int yCoordinate = -1;
        private PieceColor pieceColor;
        
        public ChessBoard ChessBoard
        {
            get { return chessBoard; }
            set { chessBoard = value; }
        }

        public int XCoordinate
        {
            get { return xCoordinate; }
            set { xCoordinate = value; }
        }
        
        public int YCoordinate
        {
            get { return yCoordinate; }
            set { yCoordinate = value; }
        }

        public PieceColor PieceColor
        {
            get { return pieceColor; }
            private set { pieceColor = value; }
        }

        public Pawn(PieceColor pieceColor)
        {
            this.pieceColor = pieceColor;
        }

        public void Move(MovementType movementType, int newX, int newY)
        {
            if (chessBoard == null || !chessBoard.IsLegalBoardPosition(newX, newY))
            {
                return;
            }
            int dx = newX - xCoordinate;
            int dy = pieceColor == PieceColor.White ?
                       newY - yCoordinate : yCoordinate - newY;
            switch (movementType)
            {
                case MovementType.Move:
                    if (dx == 0 && dy == 1 && null == chessBoard.GetPawn(newX, newY))
                    {
                        xCoordinate = newX;
                        yCoordinate = newY;
                    }
                    break;
                case MovementType.Capture:
                    Pawn p = chessBoard.GetPawn(newX, newY);
                    if (p != null)
                    {
                        PieceColor c = p.PieceColor;
                        if ((dx == 1 || dx == -1) && dy == 1 &&
                             PieceColor == c.Invert())
                        {
                            xCoordinate = newX;
                            yCoordinate = newY;
                        }
                    }
                    break;
            }
        }

        public override string ToString()
        {
            return CurrentPositionAsString();
        }

        protected string CurrentPositionAsString()
        {
            string nl = Environment.NewLine;
            return $"Current X: {XCoordinate}{nl}Current Y: {YCoordinate}{nl}Piece Color: {PieceColor}";
        }

    }
}
